src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToLinqMethodProvider.cs

```diff
@@ -0,0 +1,169 @@
+                return true;
+            }
+
+            private static ExpressionSyntax ProcessQueryBody(ConversionContext context, QueryBodySyntax queryBody)
+            {
+                var expression = context.Expression;
+                foreach (var queryClause in queryBody.Clauses)
+                {
+                    switch (queryClause.Kind())
+                    {
+                        case SyntaxKind.WhereClause: expression = CreateInvocationExpression(expression, context.Identifier, ((WhereClauseSyntax)queryClause).Condition, "Where"); break;
+                        case SyntaxKind.OrderByClause: expression = ProcessOrderByClause(expression, context.Identifier, (OrderByClauseSyntax)queryClause); break;
+                        case SyntaxKind.JoinClause: // Not supported because linq queries seem to provide essentially simpler syntax for the same than query methods.
+                        case SyntaxKind.FromClause: // More than one fromClause is not supported. The linq method seems to be more complicated for this.
+                        default: return null;
+                    }
+                }
+
+                var selectOrGroupClause = queryBody.SelectOrGroup;
+                switch (selectOrGroupClause.Kind())
+                {
+                    case SyntaxKind.SelectClause: expression = ProcesSelectClause(expression, context.Identifier, (SelectClauseSyntax)selectOrGroupClause); break;
+                    case SyntaxKind.GroupClause: expression = ProcessGroupClause(expression, context.Identifier, (GroupClauseSyntax)selectOrGroupClause); break;
+                    case SyntaxKind.LetClause: return null; // Skip this because the linq method for the let will be more complicated than the query.
+                    default: return null;
+                }
+
+                if (queryBody.Continuation != null)
+                {
+                    context = new ConversionContext(queryBody.Continuation.Identifier, expression);
+                    return ProcessQueryBody(context, queryBody.Continuation.Body);
+                }
+
+                return expression;
+            }
+
+            private static ExpressionSyntax ProcessGroupClause(ExpressionSyntax expression, SyntaxToken identifier, GroupClauseSyntax groupClause)
+            {
+                var parameter = SyntaxFactory.Parameter(identifier);
+                var groupExpressionLambda = SyntaxFactory.SimpleLambdaExpression(parameter, groupClause.GroupExpression.WithoutTrailingTrivia());
+                var byExpressionLambda = SyntaxFactory.SimpleLambdaExpression(parameter, groupClause.ByExpression.WithoutTrailingTrivia());
+                var groupExpressionArgument = SyntaxFactory.Argument(groupExpressionLambda);
+                var byExpressionArgument = SyntaxFactory.Argument(byExpressionLambda);
+                var arguments = new SeparatedSyntaxList<ArgumentSyntax>().Add(byExpressionArgument).Add(groupExpressionArgument);
```

Comment:
SyntaxFactory.SeparatedList

Reply:
would use syntax factory but still have to use Add

---
In reply to: [167387637](https://github.com/dotnet/roslyn/pull/24759#discussion_r167387637) [](ancestors = 167387637)

Reply:
why would yoy have to use Add?  just do SyntaxFactory.SeparatedList(byExpressionArgument, groupExpressionArgument) right?

Reply:
I do not see such method. Do we have it?

---
In reply to: [167747477](https://github.com/dotnet/roslyn/pull/24759#discussion_r167747477) [](ancestors = 167747477)

Reply:
Sorry, this would be: ```SyntaxFactory.SeparatedList(new [] { byExpressionArgument, groupExpressionArgument })```.  Surprised we don't have the params form for this overload.

Reply:
Thank you! I hope that the call of this method will be reasonable from performance view. Will apply it.

---
In reply to: [167748594](https://github.com/dotnet/roslyn/pull/24759#discussion_r167748594) [](ancestors = 167748594)

Reply:
A user is going to be performing this refactoring on a piece of their code.  It's virtually impossible that any reasonable existing query expression would cause this to have poor performance :)